lib/wasix/src/syscalls/wasix/call_dynamic.rs

```diff
@@ -0,0 +1,80 @@
+use wasmer::Type;
+use super::*;
+use crate::syscalls::*;
+
+/// TODO: write proper documentation for this function
+/// Calls a function from the indirect function table with the given values.
+/// 
+/// This function can be used to call functions whose types are not known at compile time of the caller.
+/// 
+/// As the caller it is you responsibility
+///  need to pass the correct number of values and with the correct types. 
+/// 
+/// function_id:
+/// The ID of the function to call
+/// This is an index into the indirect function table
+/// 
+/// values:
+/// Pointer to a sequence of pointers to the values that should be passed to the function
+/// If the function does not take any values, this can be a nullptr (0).
+/// As the user of this function, you need to ensure that the values are in matching the actual function signature. 
+/// If they dont match, its undefined behavior.
+/// 
+/// return_value:
+/// Pointer to a location where the return value of the function will be written.
+/// If the function does not return a value, this can be a nullptr (0).
+/// As the user of this function, its your responsibility to ensure that the return buffer is large enough to hold the return value.
+/// 
+#[instrument(level = "trace", skip_all, fields(path = field::Empty), ret)]
+pub fn call_dynamic<M: MemorySize>(
+    mut ctx: FunctionEnvMut<'_, WasiEnv>,
+    function_id: u32,
+    values: WasmPtr<u8, M>,
+    results: WasmPtr<u8, M>,
+) -> Result<Errno, WasiRuntimeError> {
+    WasiEnv::do_pending_operations(&mut ctx)?;
+
+    let (env, mut store) = ctx.data_and_store_mut();
+
+    let Some(indirect_function_table) = env.inner().main_module_indirect_function_table() else {
+        // No function table is available, so we cannot call any functions dynamically.
+        // TODO: This should cause a hard crash, but we return an error for now
+        return Ok(Errno::Notsup);
+    };
+    let Some(function_value) = indirect_function_table.get(&mut store, function_id) else {
+        // TODO: This should cause a trap similar to calling a function with an invalid index in the function table.
+        return Ok(Errno::Inval);
+    };
+        
+    let Value::FuncRef(Some(function)) = function_value else {
+        // The table does not contain functions, but something else.
+        // TODO: This should cause a crash 
+        return Ok(Errno::Inval);
```

Comment:
This should be caught at a much earlier stage, and be an internal error at this point. When caching the table's export in the handles struct (see my comment above) we can verify it has the correct type.